src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -70,22 +72,31 @@ public JsonClassInfo ElementClassInfo
 
         public void UpdateSortedPropertyCache(ref ReadStackFrame frame)
         {
-            // Check if we are trying to build the sorted cache.
-            if (frame.PropertyRefCache == null)
-            {
-                return;
-            }
+            Debug.Assert(frame.PropertyRefCache != null);
 
-            List<PropertyRef> newList;
-            if (_propertyRefsSorted != null)
-            {
-                newList = new List<PropertyRef>(_propertyRefsSorted);
-                newList.AddRange(frame.PropertyRefCache);
-                _propertyRefsSorted = newList.ToArray();
-            }
-            else
+            lock (_cacheLock)
             {
-                _propertyRefsSorted = frame.PropertyRefCache.ToArray();
+                if (_propertyRefsSorted != null)
+                {
+                    List<PropertyRef> replacementList = new List<PropertyRef>(_propertyRefsSorted);
+                    Debug.Assert(replacementList.Count <= PropertyNameCountCacheThreshold);
+
+                    // Verify replacementList will not become too large. This could occur in race conditions.
+                    List<PropertyRef> listToAppend = frame.PropertyRefCache;
+                    while (replacementList.Count + listToAppend.Count > PropertyNameCountCacheThreshold)
+                    {
+                        listToAppend.RemoveAt(listToAppend.Count - 1);
+                    }
+
+                    // Add the new items; we don't check for duplicates so in a race condition duplicates are possible
```

Comment:
Why not check for duplicates?  What are the pros and cons?

Reply:
Con: expensive since a temporary (or cached) dictionary would have to be used (or maintained). Although this is "warm-up" code and performance is not that critical, it would slow down this area by quite a bit (if a temporary dictionary used), or use additional global memory to maintain the dictionary (if dictionary is cached).
Pro: if duplicated prevented, there would be better performance during deserialization if duplicates appear before non-duplicates. This can only occur if JSON is not always in the same order and there is a race condition to build up the cache for the same POCO type during warm-up.